### PR TITLE
Implement Sampling-based report visualisation for Mizar

### PR DESCRIPTION
**Pull request by akopich**
This partially implements visualisation of the report
based on the sampling and frame track data. Namely, function name,
exclusive percentage and exclusive active function time per frame
are shown.

Also, this templates-up SamplingWithFrameTrackComparisonReport.

Tests: Unit, Manual
Bug: http://b/238307352

### CODE REVIEW

**Comment by akopich** [comment]
https://screenshot.googleplex.com/3YRRMSqzmssjB6u.png

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackReportModel.h [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  }
+
+  [[nodiscard]] int rowCount(const QModelIndex& /*parent*/) const override {
+    return static_cast<int>(sfids_.size());
+  };
+  [[nodiscard]] int columnCount(const QModelIndex& /*parent*/) const override {
+    return kColumnsCount;
+  };
+  [[nodiscard]] QVariant data(const QModelIndex& index, int role) const override {
+    if (role != Qt::DisplayRole) return {};
```
Comment (beckerhe) [inline src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackReportModel.h:62]: I simple sanity check for the `index` would be great. I know there is a very tight contract in place that Qt Views won't violate, but `data` is a public method and we shouldn't crash no matter what the user provides here.

I think it's enough whether the index has been created by the model:

```
if (index.model() != this) return {};
```

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackReportModel.h [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    const SFID sfid = sfids_[index.row()];
+    const auto column = static_cast<Column>(index.column());
+    switch (column) {
+      case Column::kFunctionName:
+        return report_.GetSfidToNames().at(sfid);
+      case Column::kBaselineExclusivePercent:
+      case Column::kBaselineExclusiveTimePerFrame:
+      case Column::kComparisonExclusivePercent:
+      case Column::kComparisonExclusiveTimePerFrame:
+        return absl::StrFormat("%.3f", (MakeNumericEntry(sfid, column)));
```
Comment (beckerhe) [inline src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackReportModel.h:97]: What's the rationale for having `MakeNumericEntry`? Is it to avoid having multiple `absl::StrFormat` calls?

Comment (beckerhe) [reply src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackReportModel.h:97]: If that was the only reason I would argue that it would be more readable to have redundant StrFormat calls over the separate function - mainly because the Column enum makes it hard to distinguish which cases need to be handled in which function and it ties the two function together anyway.

File: src/MizarWidgets/SamplingWithFrameTrackReportModelTest.cpp [hunk 3]
```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <absl/container/flat_hash_map.h>
+#include <absl/container/flat_hash_set.h>
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <qnamespace.h>
```
Comment (beckerhe) [inline src/MizarWidgets/SamplingWithFrameTrackReportModelTest.cpp:9]: Nit:  :-)

File: src/MizarWidgets/SamplingWithFrameTrackReportModelTest.cpp [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                 baseline_frame_track_stats_,
+                 comparison_counts_,
+                 comparison_frame_track_stats_,
+                 {},
+                 &kSfidToName};
+
+  Model model_{report_};
+};
+
+double ExpectDoubleAndParse(const QString& str) {
```
Comment (beckerhe) [inline src/MizarWidgets/SamplingWithFrameTrackReportModelTest.cpp:115]: static?

File: src/MizarWidgets/SamplingWithFrameTrackReportModelTest.cpp [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+      MakeBaseline<MockCounts>(kCallstacksCount, kSfidToBaselineCounts);
+  Comparison<MockCounts> comparison_counts_ =
+      MakeComparison<MockCounts>(kCallstacksCount, kSfidToComparisonCounts);
+
+  Baseline<MockFrameTrackStats> baseline_frame_track_stats_ =
+      MakeBaseline<MockFrameTrackStats>(kBaselineFrameTime);
+  Comparison<MockFrameTrackStats> comparison_frame_track_stats_ =
+      MakeComparison<MockFrameTrackStats>(kComparisonFrameTime);
+
+  QString DisplayedString(int row, Column column) {
```
Comment (beckerhe) [inline src/MizarWidgets/SamplingWithFrameTrackReportModelTest.cpp:100]: const?

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackReportModel.h [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+  [[nodiscard]] int columnCount(const QModelIndex& /*parent*/) const override {
+    return kColumnsCount;
+  };
+  [[nodiscard]] QVariant data(const QModelIndex& index, int role) const override {
+    if (role != Qt::DisplayRole) return {};
+    return QString::fromStdString(MakeDisplayedString(index));
+  }
+
+  [[nodiscard]] QVariant headerData(int section, Qt::Orientation orientation,
+                                    int role) const override {
```
Comment (beckerhe) [inline src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackReportModel.h:67]: All these model functions have default arguments in Qt. (Like `QModelIndex{}` for the index in `rowCount`.)

We don't want to have default arguments for virtual functions because they can get out of sync - like here. Could you sync them up since we can't change them in Qt?

File: src/MizarData/include/MizarData/SamplingWithFrameTrackComparisonReport.h [hunk 7]
```diff
@@ -85,55 +86,66 @@ class SamplingCounts {
...
-  Comparison<orbit_client_data::ScopeStats> comparison_frame_track_stats_;
+  Comparison<Counts> comparison_sampling_counts_;
+  Comparison<FrameTrackStats> comparison_frame_track_stats_;
 
   absl::flat_hash_map<SFID, CorrectedComparisonResult> fid_to_corrected_comparison_results_;
+
+  const absl::flat_hash_map<SFID, std::string>* sfid_to_names_;
 };
 
+using SamplingWithFrameTrackComparisonReport =
```
Comment (florian-kuebler) [inline src/MizarData/include/MizarData/SamplingWithFrameTrackComparisonReport.h:146]: As per [style guide](https://google.github.io/styleguide/cppguide.html#Aliases) for aliases in the public API:

>When defining a public alias, document the intent of the new name, including whether it is guaranteed to always be the same as the type it's currently aliased to, or whether a more limited compatibility is intended.

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackOutputWidget.h [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+#include "MizarData/SamplingWithFrameTrackComparisonReport.h"
+
+namespace Ui {
+class SamplingWithFrameTrackOutputWidget;
+}
+
+namespace orbit_mizar_widgets {
+
+class SamplingWithFrameTrackOutputWidget : public QWidget {
```
Comment (florian-kuebler) [inline src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackOutputWidget.h:18]: Please add a class level comment.

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackOutputWidget.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  using Report = ::orbit_mizar_data::SamplingWithFrameTrackComparisonReport;
```
Comment (florian-kuebler) [inline src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackOutputWidget.h:19]: Help me with C++ scoping again. So this alias is only within this class, right?

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackReportModel.h [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+#include <vector>
+
+#include "MizarData/SamplingWithFrameTrackComparisonReport.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/Typedef.h"
+
+namespace orbit_mizar_widgets {
+
+template <typename Report, typename Counts, typename FrameTrackStats>
+class SamplingWithFrameTrackReportModelTmpl : public QAbstractTableModel {
```
Comment (florian-kuebler) [inline src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackReportModel.h:24]: Again, please add a class level comment.

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackWidget.h [hunk 11]
```diff
@@ -32,6 +32,8 @@ class SamplingWithFrameTrackWidget : public QWidget {
   template <typename T>
   using Comparison = ::orbit_mizar_base::Comparison<T>;
 
+  using Report = ::orbit_mizar_data::SamplingWithFrameTrackComparisonReport;
```
Comment (florian-kuebler) [inline src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackWidget.h:35]: Why is this needed?

**Review by florian-kuebler — APPROVED** [review]
Only some minor comments. LGTM.

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackOutputWidget.h (hunk 9 above)
Comment (akopich) [reply src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackOutputWidget.h:19]: Yes, and it is private.

File: src/MizarWidgets/SamplingWithFrameTrackOutputWidget.cpp [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+SamplingWithFrameTrackOutputWidget::SamplingWithFrameTrackOutputWidget(QWidget* parent)
+    : QWidget(parent), ui_(std::make_unique<Ui::SamplingWithFrameTrackOutputWidget>()) {
+  ui_->setupUi(this);
+}
+
+void SamplingWithFrameTrackOutputWidget::UpdateReport(Report report) {
+  auto model = std::make_unique<SamplingWithFrameTrackReportModel>(std::move(report), this);
+
+  ui_->report_->setModel(model.release());
+  ResizeReportColumns(width());
```
[the remark this answers is not in the document]
Comment (akopich) [reply src/MizarWidgets/SamplingWithFrameTrackOutputWidget.cpp:27]: The report is driven by the choice of threads and frame track. Yeah, I am afraid the change of such yields a completely new report. 

But if the input changes encompasses only the change of the start fields, you're right. But I plan to remove the start field as soon as the kind of functionality is supported in Orbit itself.

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackWidget.h (hunk 11 above)
Comment (akopich) [reply src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackWidget.h:35]: This is needed only in the .cpp file. So you're right, it's out of place here. Moved to the cpp file.
